### ISSUE TITLE
Canvas Kit React API Change Proposal

### ISSUE THREAD

**Issue by anicholls** [Workday/canvas-kit#51]
As part of the 3.0 Alpha, we are making some breaking changes to solidify our APIs, particularly in accordance with our [API & Pattern Guidelines](https://github.com/Workday/canvas-kit/blob/master/API_PATTERN_GUIDELINES.md). 

The following are proposed API changes. This should probably move to our RFC repo once a public one is created.

# API Changes

## Universal Changes
- (https://github.com/Workday/canvas-kit/issues/95 / https://github.com/Workday/canvas-kit/pull/150) ~~**Not Breaking**: Rename `props` (ActionBar), `inputProps` (TextArea, TextInput) and `otherProps` to `elemProps`~~

- ~~**Maybe Breaking** (https://github.com/Workday/canvas-kit/issues/162) Change `HTMLAttributes` usage to `HTMLProps`~~ @lychyi 

- ~~Add a beta/sandbox/unstable module~~ @anicholls

### Avatar

- (https://github.com/Workday/canvas-kit/pull/194) ~~Change `themeColor` to `variant`~~ @stephanerangaya 

### Banner 

- (https://github.com/Workday/canvas-kit/issues/96 / https://github.com/Workday/canvas-kit/pull/110) ~~**Not Breaking**: Export enums as static variables~~
- (https://github.com/Workday/canvas-kit/pull/195) ~~Change `theme` to `variant`.~~ @stephanerangaya 

### ~~Button~~

- (https://github.com/Workday/canvas-kit/issues/97 / https://github.com/Workday/canvas-kit/pull/129) ~~Rename `buttonType` to `type`~~
- (https://github.com/Workday/canvas-kit/issues/97 / https://github.com/Workday/canvas-kit/pull/129) ~~Rename `buttonSize` to `size`~~
- (https://github.com/Workday/canvas-kit/issues/97 / https://github.com/Workday/canvas-kit/pull/129) ~~Rename all types to singular~~
- (https://github.com/Workday/canvas-kit/issues/97 / https://github.com/Workday/canvas-kit/pull/129) ~~Rename all static variables to singular~~
- (https://github.com/Workday/canvas-kit/issues/97 / https://github.com/Workday/canvas-kit/pull/129) ~~Re-evaluate prop spread behavior~~
- (https://github.com/Workday/canvas-kit/issues/97 / https://github.com/Workday/canvas-kit/pull/129) ~~Make aria-label required for IconButton~~
- (https://github.com/Workday/canvas-kit/issues/97 / https://github.com/Workday/canvas-kit/pull/129) ~~Consider scaling down `Small` `SquareFilled` Icon Button icon to match the other `Small` icon~~
- (https://github.com/Workday/canvas-kit/issues/97 / https://github.com/Workday/canvas-kit/pull/129) ~~Fix the focus outline on square buttons to match others (it's the only one that doesn't have the white ring inside the blue) (see https://github.com/Workday/canvas-kit/pull/53#discussion_r312263943)~~
- ~~Promoting `beta_Button` to `Button` and deprecating `Button`~~ @mannycarrera4 

### Card

- (https://github.com/Workday/canvas-kit/issues/98) / (https://github.com/Workday/canvas-kit/pull/204) ~~Update border radius (ripples to Popup, etc)~~ @stephanerangaya  

### ~~ColorPicker~~

- (https://github.com/Workday/canvas-kit/issues/99 / https://github.com/Workday/canvas-kit/pull/150) ~~**Not Breaking**: For `ColorPicker`, pass `elemProps` to `ColorPreviewComponent`~~

### Core

- ~~Move `beta_type` to sandbox~~ @sahlhoff (blocked by @anicholls PR)
- ~~Make spacing enums singular~~ @mannycarrera4 

### ~~Form Field~~

- (https://github.com/Workday/canvas-kit/issues/100 / https://github.com/Workday/canvas-kit/pull/146) ~~Should `LabelPosition`, `ErrorBehavior`, etc. be prefixed with component name? We do this elsewhere~~
- (https://github.com/Workday/canvas-kit/issues/99 /  https://github.com/Workday/canvas-kit/pull/150) ~~Add `elemProps`~~

### Header

- (https://github.com/Workday/canvas-kit/pull/157) Split search out into a combobox component 
- Header needs a larger refactor. Will come back to this.
- ~~Move to unstable/sandbox module~~ @sahlhoff  (blocked by @anicholls sandbox PR and @vibdev combox PR)

### ~~Layout~~

- (https://github.com/Workday/canvas-kit/issues/99 / #150 ) ~~**Not Breaking**: Unpack props for both components~~

### ~~Loading-animation~~

- (https://github.com/Workday/canvas-kit/issues/103 / https://github.com/Workday/canvas-kit/pull/142) ~~Deprecate loading spinner~~
- (https://github.com/Workday/canvas-kit/issues/99 / https://github.com/Workday/canvas-kit/pull/150) ~~**Not Breaking**: Add `elemProps`~~

### Menu

- Use `Popup` so you don't have to wrap once we have a11y fixes @NicholasBoll 
- ~~Move to unstable/sandbox~~ @sahlhoff  (blocked by @anicholls PR)

### ~~Modal~~

- (https://github.com/Workday/canvas-kit/issues/99 / https://github.com/Workday/canvas-kit/pull/150) ~~**Not Breaking**: Add `elemProps`~~
- ~~Remove `transformOrigin` prop. Modals should always animate from the bottom up~~ @NicholasBoll 

### ~~Page Header~~

- (https://github.com/Workday/canvas-kit/issues/104 / https://github.com/Workday/canvas-kit/pull/138) ~~Change `marketing` prop~~

### Popup

- (https://github.com/Workday/canvas-kit/pull/110) ~~Capitalize `Popup.padding` static variable~~
- (https://github.com/Workday/canvas-kit/issues/59) ~~A11y updates~~ @NicholasBoll 

### ~~SidePanel~~

- (https://github.com/Workday/canvas-kit/issues/105 / https://github.com/Workday/canvas-kit/pull/123) ~~Change `sidePanelBackgroundColor` to `backgroundColor`~~
- (https://github.com/Workday/canvas-kit/issues/105 / https://github.com/Workday/canvas-kit/pull/123) ~~**Not Breaking**: add `backgroundColor` as a static variable~~
- (https://github.com/Workday/canvas-kit/issues/105 / https://github.com/Workday/canvas-kit/pull/123) ~~**Not Breaking**: Lots of broken styles (`undefinedpx`), etc.~~
- (https://github.com/Workday/canvas-kit/issues/105 / https://github.com/Workday/canvas-kit/pull/123) ~~**Not Breaking:** Change `open` param in `onBreakingChange` to `isAboveBreakpoint` or something similar. We shouldn't suggest the behavior.~~

### ~~Skeleton~~

- (https://github.com/Workday/canvas-kit/issues/99 / https://github.com/Workday/canvas-kit/pull/150) ~~**Not Breaking**: Add `elemProps`?~~

### Switch

- (https://github.com/Workday/canvas-kit/pull/197) ~~**Not Breaking**: Add a11y messaging in Readme (the same as checkbox)~~ @stephanerangaya 

### ~~Table~~

- https://github.com/Workday/canvas-kit/pull/172 ~~TableRow~~
  - ~~Singularize `TableRowStates` @sahlhoff~~
  - ~~Update to use common error/alert enums and colors @sahlhoff~~
  - (https://github.com/Workday/canvas-kit/issues/99 / https://github.com/Workday/canvas-kit/pull/150) ~~**NOT BREAKING**: Add `elemProps`~~

**Comment by anicholls** [comment]
Question: Should we be renaming tshirt sizes to capitals? Technically they're not [PascalCase](https://github.com/Workday/canvas-kit/blob/master/API_PATTERN_GUIDELINES.md#enums)

**Comment by anicholls** [comment]
The following items have been postponed:



## Universal Changes

- [POSTPONED] (https://github.com/Workday/canvas-kit/issues/94) **Not Breaking**: Add support for `Box` component wrapper. We should probably standardize the variable name too (i.e. `Container`)

- [POSTPONED] Consider breakpoint system across all components

- [POSTPONED] **Not Breaking**: Do we extend `React.HTMLAttributes<HTMLDivElement>` for basic div containers? If so, a few components need it (RadioGroup, Toast, Tooltip). How does this interact with `Box` wrapper? Related to https://github.com/Workday/canvas-kit/pull/150#issuecomment-525877057 - let's leave for now.

- [POSTPONED] **Not Breaking**: Audit for hard coded type that we can switch to core type. @sahlhoff (will look at linting as well)

- [POSTPONED] (https://github.com/Workday/canvas-kit/issues/58) **Not Breaking**: Add border radius variable @mannycarrera4 

- Consider using objects rather than enums for some types. @NicholasBoll 

  - We could probably accomplish the same thing with an object instead and then `ButtonTypes.Primary` would also work in JavaScript (Instead of just in TypeScript). Also TypeScript would then allow hard-coding `'primary'` like is already allowed if using JavaScript. Ex:

    ```
    export const ButtonTypes = {
      Primary: 'primary',
      Secondary: 'secondary',
      Delete: 'delete',
    } as const;
    
    // "primary" | "secondary" | "delete"
    type ButtonTypeValues =  typeof ButtonTypes[keyof typeof ButtonTypes]
    ```
    The `as const` forces the values of `ButtonTypes` to be constrained to string literals instead of `string`
    The implementation looks a bit more tricky, but is more consistent on the consumer side. It allows you to use the string literals (i.e. `<Button type="primary" />`)

- [POSTPONED - Theming]  Update any color types to `CanvasColor` from `string` 

- [POSTPONED] **Not Breaking**: Change all [defaultProps to required in the interfaces](https://github.com/Workday/canvas-kit/blob/master/API_PATTERN_GUIDELINES.md#default-props)  (maybe add a linter?)

### Core

- [POSTPONED] Make `beta_type` default now. Update all other components to match. In migration docs, specify that they should use 
- [POSTPONED] Pull design-assets over

### Icon

- [POSTPONED] (https://github.com/Workday/canvas-kit/issues/101) **Not Breaking** (because we don't export `Icon`): Update wrapper classes to spread `elemProps` instead of passing `elemProps`
- [POSTPONED] Remove `elemProps` from `Svg`



### Menu

- [POSTPONED] ~~Review Rob's PR and decide what we want to do.~~ @NicholasBoll 



### Skeleton

- [POSTPONED] (https://github.com/Workday/canvas-kit/issues/134) **Not Breaking**: Audit animation 
- [POSTPONED] **Not Breaking**: rename `utils.js` file
- [POSTPONED] **Not Breaking**: Update shape styled div to use interface



### Toast

- [POSTPONED - Theming] Change `iconColor` to only accept `CanvasColor`

**Comment by anicholls** [comment]
All of the changes required for 3.0 tracked here have made it in for the most part. Going to close this for now and track the remainder as separate issues.
